pycozmo/audio.py

```diff
@@ -0,0 +1,147 @@
+import struct
+import time
+import wave
+from datetime import datetime, timedelta
+from threading import Thread, Lock
+
+from .conn import ClientConnection
+from .protocol_encoder import OutputAudio, Keyframe
+
+MIN_WAIT = 0.033
+
+class AudioManager():
+    """
+    This class takes care of reading audio files and generating the OutputAudio messages sent to
+    cozmo.
+
+    The play() method can be used to play an audio file or list of OutputAudio messages.
+
+    Args:
+        conn (ClientConnection): client managing the communication with the robot
+    """
+    def __init__(self, conn: ClientConnection):
+        self.stream = []
+        self.stop = False
+        self.conn = conn
+        self.thread = Thread()
+        self.lock = Lock()
+        self.audio_gen = []
+        self.conn.add_handler(Keyframe, self._on_keyframe)
+
+    def start(self):
+        self.stop = False
+        if not self.thread.is_alive():
+            self.thread = Thread(target=self.run, name=__class__.__name__)
+            self.thread.start()
+
+    def stop(self) -> None:
+        self.stop = True
+        self.thread.join()
+        self.audio_gen = []
+
+    def run(self) -> None:
+        while len(self.audio_gen) > 0:
+            generator = []
+
+            with self.lock:
+                generator = self.audio_gen.pop(0)
```

Comment:
So, this was classic me experimenting a bit with the language. The idea behind it is that instead of having to load the whole file, you can load frame by frame on the go. I thought this would be nice because then you don't need to do all the processing of the file before starting the audio. You could also argue that this will consume less memory, but unless you are loading a big file, I don't think it will matter.

Also, I think you would only see the difference when the thread is not yet running. If there is already something playing this will run in a separate thread, so it should not be a problem.

I don't mind having the generators or the list preloaded, what would you prefer?

Reply:
The current solution is a good start.

As animation playback with Cozmo is very time critical, in the long run, I think we should switch to preloading and preprocessing everything so that thread creation and I/O do not interfere with playback.

Reply:
That sounds good to me, although I'm not sure I understand what would be different from a threading standpoint. I also thought that since the messages need to be sent out at regular intervals (cozmo doesn't really take care of requesting new frames, instead it keeps overloading its audio buffer), it should be enough time for the reading of the files.

Thinking about the playback,  what about including two play methods, one appending frames to the playback list, and another one clearing it and including the new data? Right now it is thought to append new audio at the end of the audio stream, but there is no option to stop the reproduction and start the new one. This might be good for the synchronization also, what do you think? 

In any case, I'm happy to change it to a preloading, will upload it in the next commit.